romanisim/l3.py

```diff
@@ -24,100 +24,167 @@
+    Source injection into an L3 image relies on knowing the objects'
+    x and y locations, the PSF, and the exposure time; if these are not
+    provided, reasonable defaults are generated.
+
+    The simulation proceeds by (optionally) using the model WCS to generate the
+    x & y locations.  It also optionally computes an appropriate pixel-convolved
+    PSF for the image.  It optionally uses the model Poisson variances and
+    model fluxes to estimate the per-source exposure times appropriate
+    for injecting the sources in the catalog onto the image.  Finally,
+    it updates the var_poisson of the input image to account for the additional
+    variance from the added sources.
+
+    Parameters
+    ----------
+    model: roman_datamodels.datamodels.WfiMosaic
+        model into which to inject sources
+    cat: astropy.table.Table
+        catalog of sources to inject into image
+    x: list[float] or None
+        x coordinates of catalog locations in image
+    y: list[float] or None
+        y coordinates of catalog locations in image
+    exptimes : list[float] or None
+        exposure times of each source.  Computed from model.var_poisson
+        and model.flux at each source location if not specified.
+    psf: galsim.gsobject.GSObject
+        PSF to use
+    rng: galsim.BaseDeviate
+        galsim random number generator to use
+    seed : int
+        Seed to use for rng
+    webbpsf: bool
+        if True, use WebbPSF to model the PSF
+
+    Returns
+    -------
+    outinfo : np.ndarray with information about added sources
     """
-    pass
+    if seed is None:
+        seed = 125
+    if rng is None:
+        rng = galsim.UniformDeviate(seed)
+
+    if x is None or y is None:
+        x, y = model.meta.wcs.numerical_inverse(cat['ra'], cat['dec'],
+                                                with_bounding_box=False)
+
+    filter_name = model.meta.basic.optical_element
+    cat = romanisim.catalog.table_to_catalog(cat, [filter_name])
+
+    wcs = romanisim.wcs.GWCS(model.meta.wcs)
+    if psf is None:
+        pixscalefrac = get_pixscalefrac(wcs, model.data.shape)
+        if (pixscalefrac > 1) or (pixscalefrac < 0):
+            raise ValueError('weird pixscale!')
+        psf = l3_psf(filter_name, pixscalefrac, webbpsf=True, chromatic=False)
+
+    maggytoes = romanisim.bandpass.get_abflux(filter_name)
+    etomjysr = romanisim.bandpass.etomjysr(filter_name)
+
+    Ct = []
+    for idx, (x0, y0) in enumerate(zip(x, y)):
+        # Set scaling factor for injected sources
+        # Flux / sigma_p^2
+        xidx, yidx = int(np.round(x0)), int(np.round(y0))
+        if model.var_poisson[yidx, xidx].value != 0:
+            Ct.append(math.fabs(
+                model.data[yidx, xidx].value /
+                model.var_poisson[yidx, xidx].value))
+        else:
+            Ct.append(1.0)
+    Ct = np.array(Ct)
+    # etomjysr = 1/C; C converts fluxes to electrons
+    exptimes = Ct * etomjysr
```

Comment:
I am concerned about going e/s to s without gain (vs. cps to s), maybe there is a cancellation I am missing. obviously, this is irrelevant when gain is 1.

Reply:
Ah.  No, I use count and electron as synonymous, and DN as something else (counts or electrons divided by the gain).  The thing that is important about counts / electrons is that they're the thing that is Poisson distributed.  The special thing about DN is that the read noise is specified in those units; otherwise we wouldn't need DN at all in this module.